Tensile/SolutionStructs.py

```diff
@@ -3582,57 +3582,65 @@ def assignDerivedParameters(state):
-      state["NoTailLoop"] = True
-    elif state["PersistentKernel"] and state["PrefetchAcrossPersistent"] and state["PrefetchAcrossPersistentMode"] == 1:
-      if not validNoTailLoop:
-        reject(state, "PAP + PAPMode + (AssertSummationElementMultiple/GlobalSplitU)%%DepthU!=0 %s to enable NoTailLoop"%invalidComment)
-        return
-      else:
-        state["NoTailLoop"] = True
-    elif state["DirectToVgprA"] or state["DirectToVgprB"]:
-      # use noTailLoop only when it is valid (NT only)
-      # otherwise, use tail loop (no reject)
-      if validNoTailLoop:
-        state["NoTailLoop"] = True
+    asemDivGSU = 1 if asem%gsu !=0 else asem//gsu
+
+    noTailLoop = 0
+    if (asemDivGSU % state["DepthU"] == 0):
+      noTailLoop = 1
+
+    # reject conditions for noTailLoop==0
+    rejected = False
+    rejectMessage = ""
+    if state["PersistentKernel"] and state["PrefetchAcrossPersistent"] and state["PrefetchAcrossPersistentMode"] == 1:
+      rejectMessage = "PK + PAP + PAPMode"
+      rejected = True
     elif state["DirectToLds"]:
+      if (not rejected) and state["EnableMatrixInstruction"] and state["LocalReadVectorWidth"] > state["MIInputPerThread"]:
+        rejectMessage = "DirectToLds + LocalReadVectorWidth>MIInputPerThread"
+        rejected = True
       for tc in ('A','B'):
-        if state["ProblemType"]["TLU%c"%tc] and state["NumLoadsCoalesced%c"%tc] > 1:
-          if not validNoTailLoop:
-            reject(state, "DirectToLds + TLU + NumLoadsCoalesced>1 + (AssertSummationElementMultiple/GlobalSplitU)%%DepthU!=0 %s to enable NoTailLoop"%invalidComment)
-            return
-          else:
-            state["NoTailLoop"] = True
-        elif state["EnableMatrixInstruction"] and state["LocalReadVectorWidth"] > state["MIInputPerThread"]:
-          if not validNoTailLoop:
-            reject(state, "DirectToLds + LocalReadVectorWidth>MIInputPerThread + (AssertSummationElementMultiple/GlobalSplitU)%%DepthU!=0 %s to enable NoTailLoop"%invalidComment)
-            return
-          else:
-            state["NoTailLoop"] = True
+        if (not rejected) and state["ProblemType"]["TLU%c"%tc] and state["NumLoadsCoalesced%c"%tc] > 1:
+          rejectMessage = "DirectToLds + TLU%c + NumLoadsCoalesced%c>1"%(tc, tc)
+          rejected = True
+
+    if noTailLoop == 0 and rejected:
+      # if reject condition for NoTailLoop is true and NoTailLoop is not enabled, reject this kernel
+      rejectMessage += " requires NoTailLoop."
+      rejectMessage += "\n" + "To enable NoTailLoop, "
+      rejectMessage += "\n" + " - AssertSummationElementMultiple/GlobalSplitU) is multiple of DepthU or"
+      rejectMessage += "\n" + " - BufferLoad and MatrixInstruction + MatrixInstK > 1 and"
+      rejectMessage += "\n" + "   (global read width for TailLoop decided by assert is multiple of GlobalReadVectorWidth) and"
+      rejectMessage += "\n" + "   (StaggerU = 0 or NT(+BufferLoad))"
+      reject(state, rejectMessage)
+      return
```

Comment:
Have you run these changes against rocblas tests to make sure current kernels are compatible with the new requirements?

Reply:
This part will not be executed with rocblas.
This new condition is basically wider than before.
Tensile kernels which was previously valid should be valid with this condition.
Regarding noTailLoop condition check, since I moved the check code in KernelWriter.py, the new logic will be applied to all existing kernels.

Reply:
Yea, that's why I'm wondering if rocblas gemm tests pass with the change. Since createLibrary is not going to hit the reject case, but it looks like old configs may get generated with new noTailLoop code?

Reply:
Thanks! Looks good!